src/dc/nipals.js

```diff
@@ -0,0 +1,71 @@
+
+/**
+ * NIPALS
+ * @param {Matrix} X
+ * @param {*} Y
+ * @param {*} u
+ *  Geladi, P and Kowalski, B.R. (1986)
+ * Partial least squares and regression:
+ * a tutorial.
+ * Analytica Chimica Acta 185, 1-17.
+ */
+export default class nipals {
+  constructor(X, options = {}) {
+    const { Y } = options;
+
+    let u;
+    if (Y) {
+      u = Y.getColumnVector(0);
+    } else {
+      u = X.getColumnVector(0);
+    }
+
+    let diff = 1;
+    let t, q, w, tOld;
+
+    let counter = 0;
+    for (var i = 0; diff > 1e-10; i++) {
```

Comment:
you don't actually use this `i` variable. It seems to me that you should instead increment the counter here.

`for (var counter = 0; diff > 1e-10 && counter <1000; counter++)`

Reply:
the conditions for the loop have been modified accordingly